### PR TITLE
[OneDNN] Optimize unsorted segment sum on 1 dim

### PR DESCRIPTION
**Pull request by kmagiers**
Optimize unsorted segment sum calculations on 1-dim by removing tensor class overheads.

Performance numbers:
| Reduction | Num rows | Num cols | Segment size | Before (ns) | After (ns) | Speed up |
| --------- | -------- | -------- | ------------ | ----------- | ---------- | -------- |
| Sum       | 4,096    | 1        | 128          | 60,812      | 5,899      | 10.31x    |
| Max       | 4,096    | 1        | 128          | 52,164      | 5,524      | 9.44x     |
| Min       | 4,096    | 1        | 128          | 52,052      | 5,957      | 8.74x     |
| Prod      | 4,096    | 1        | 128          | 66,103      | 9,573      | 6.91x     |

### CODE REVIEW

**Comment by gbaned** [comment]
Hi @kmagiers Can you please sign CLA. Thank you!

**Review by Aayog — APPROVED** [review]

**Comment by kmagiers** [comment]
> Hi @kmagiers Can you please sign CLA. Thank you!

CLA signed

**Comment by gbaned** [comment]
Hi @jsmeredith Can you please review this PR ? Thank you!

**Comment by gbaned** [comment]
Hi @jsmeredith Can you please review this PR ? Thank you!

File: tensorflow/core/kernels/segment_reduction_ops_impl.h [hunk 1]
```diff
@@ -358,8 +358,12 @@ struct UnsortedSegmentFunctor<CPUDevice, T, Index, InitialValueF, ReductionF> {
     const int64_t N = segment_ids.dimension(0);
     const int64_t num_segments = output.dimension(0);
     const int64_t inner_dim = data.dimension(1);
+    const T* data_ptr = data.data();
+    T* out_ptr = output.data();
     ReductionF reduction;
 
+    bool data_is_1D = data.dimensions()[1] == 1;
```
Comment (ishark) [inline tensorflow/core/kernels/segment_reduction_ops_impl.h:365]: nit: bool is_inner_dim_1d =  inner_dim == 1?

Comment (AnetaKaczynska) [reply tensorflow/core/kernels/segment_reduction_ops_impl.h]: done

**Comment by tbaker2** [comment]
@penpornk Aneta has run the failed test on her local system and they all pass. The tests timed out on the CI server. It looks like the CI server is too busy. How should we get this unblocked?

**Comment by gbaned** [comment]
Hi @penpornk Can you please review this PR ? Thank you!

**Comment by gbaned** [comment]
Hi @penpornk Can you please review this PR ? Thank you!

**Comment by gbaned** [comment]
Hi @penpornk Can you please review this PR ? Thank you!

**Comment by gbaned** [comment]
Hi @penpornk Can you please review this PR ? Thank you!

**Comment by gbaned** [comment]
Hi @penpornk Can you please review this PR ? Thank you!

File: tensorflow/core/kernels/segment_reduction_ops_impl.h [hunk 2]
```diff
@@ -358,8 +358,12 @@ struct UnsortedSegmentFunctor<CPUDevice, T, Index, InitialValueF, ReductionF> {
     const int64_t N = segment_ids.dimension(0);
     const int64_t num_segments = output.dimension(0);
     const int64_t inner_dim = data.dimension(1);
+    const T* data_ptr = data.data();
+    T* out_ptr = output.data();
     ReductionF reduction;
 
+    bool is_inner_dim_1d = inner_dim == 1;
```
Comment (penpornk) [inline tensorflow/core/kernels/segment_reduction_ops_impl.h:365]: Nit:
```suggestion
    const bool is_inner_dim_1d = inner_dim == 1;
```

File: tensorflow/core/kernels/segment_reduction_ops_impl.h [hunk 3]
```diff
@@ -410,15 +414,29 @@ struct UnsortedSegmentFunctor<CPUDevice, T, Index, InitialValueF, ReductionF> {
...
     const int64_t input_bytes = sizeof(T) * inner_dim * kAverTaskSize;
     const int64_t output_bytes = sizeof(T) * inner_dim * kAverTaskSize;
     const Eigen::TensorOpCost cost(input_bytes, output_bytes, compute_cycles);
-    cpu_device.parallelFor(num_segments, cost, reductionWorker);
+    if(is_inner_dim_1d) {
+        cpu_device.parallelFor(num_segments, cost, reductionWorker1D);
+    }
+    else {
+        cpu_device.parallelFor(num_segments, cost, reductionWorker);
+    }
```
Comment (penpornk) [inline tensorflow/core/kernels/segment_reduction_ops_impl.h:438]: Nit: Please apply code formatter. The formatted code will probably look like:
```suggestion
    if (is_inner_dim_1d) {
      cpu_device.parallelFor(num_segments, cost, reductionWorker1D);
    } else {
      cpu_device.parallelFor(num_segments, cost, reductionWorker);
    }
```

Alternatively, we could also do
```
    cpu_device.parallelFor(
        num_segments, cost,
        is_inner_dim_1d ? reductionWorker1D : reductionWorker);
```

**Review by penpornk — CHANGES_REQUESTED** [review]
Thank you for the PR and sorry for the delay! Could you please post some performance number comparisons? Are the speedup from avoiding Eigen's indexing overhead? 

@cantonios Would specializing Eigen indexing for 1D case be helpful?

File: tensorflow/core/kernels/segment_reduction_ops_impl.h (hunk 3 above)
Comment (cantonios) [reply tensorflow/core/kernels/segment_reduction_ops_impl.h:438]: The types are different, so you can't combine them in a ternary expression.

**Comment by cantonios** [comment]
> Thank you for the PR and sorry for the delay! Could you please post some performance number comparisons? Are the speedup from avoiding Eigen's indexing overhead?
> 
> @cantonios Would specializing Eigen indexing for 1D case be helpful?

I don't think it's just the indexing - there will be small overhead with `cwiseMax`,  tensor cwise sum, etc.  But yeah, it would be nice to have some numbers to support this.

**Review by Aayog — APPROVED** [review]

**Comment by cantonios** [comment]
> @penpornk @cantonios Here are some performance numbers:
> 
> Reduction	Num rows	Num cols	Segment size	Before (ns)	After (ns)	Speed up
> Sum	4,096	1	128	60,812	5,899	10.31x
> Max	4,096	1	128	52,164	5,524	9.44x
> Min	4,096	1	128	52,052	5,957	8.74x
> Prod	4,096	1	128	66,103	9,573	6.91x

Can you put these numbers up in the PR description so they are added to the commit message?

**Comment by AnetaKaczynska** [comment]
> > @penpornk @cantonios Here are some performance numbers:
> > Reduction	Num rows	Num cols	Segment size	Before (ns)	After (ns)	Speed up
> > Sum	4,096	1	128	60,812	5,899	10.31x
> > Max	4,096	1	128	52,164	5,524	9.44x
> > Min	4,096	1	128	52,052	5,957	8.74x
> > Prod	4,096	1	128	66,103	9,573	6.91x
> 
> Can you put these numbers up in the PR description so they are added to the commit message?

This PR was submitted by our ex-employee. I don't think I can edit his comment (issue description). Please let me know if there's some other way.

**Review by penpornk — APPROVED** [review]
Thank you very much for the numbers and fixes!

**Pull request merged**